### PR TITLE
find users for background scan one by one

### PR DESCRIPTION
**Pull request by icewind1991**
The current query for finding a list of users to trigger the background scan for has performance issue on large instances due to the grouping/deduplicating in the query.

This changes the logic to find a single user to scan, run the scan, and find a new user to scan.

Some additional work in how background scanning was handled for shared storages and other jails was needed to ensure this doesn't get stuck on a single user.

As an additional bonus, this improves the situation where a storage with unscanned files is accessible for multiple users. Currently the background scan would be triggered for all users with access, even if there is no longer anything to scan after the first user.
With these changes it will only trigger the scan for one user.

### CODE REVIEW

File: apps/files/lib/BackgroundJob/ScanFiles.php [hunk 1]
```diff
@@ -94,20 +90,20 @@ protected function runScanner(string $user) {
...
+	private function getUserToScan() {
 		$query = $this->connection->getQueryBuilder();
-		$query->select($query->func()->max('user_id'))
+		$query->select('user_id')
 			->from('filecache', 'f')
 			->innerJoin('f', 'mounts', 'm', $query->expr()->eq('storage_id', 'storage'))
 			->where($query->expr()->lt('size', $query->createNamedParameter(0, IQueryBuilder::PARAM_INT)))
-			->groupBy('storage_id')
-			->setMaxResults(self::USERS_PER_SESSION);
+			->andWhere($query->expr()->gt('parent', $query->createNamedParameter(-1, IQueryBuilder::PARAM_INT)))
```
Comment (PVince81) [inline apps/files/lib/BackgroundJob/ScanFiles.php:103]: is this reliable ? can it happen that suddenly there's a new "size=-1" folder appearing on ext storage and that it's not bubbled up to the storage root ?

Comment (PVince81) [reply apps/files/lib/BackgroundJob/ScanFiles.php:103]: wait, I read the condition in reverse. so we take all entries that are NOT the root

File: apps/files/lib/BackgroundJob/ScanFiles.php [hunk 2]
```diff
@@ -119,10 +115,12 @@ protected function run($argument) {
...
 		}
 
-		$users = $this->getUsersToScan();
-
-		foreach ($users as $user) {
+		$usersScanned = 0;
+		$user = $this->getUserToScan();
+		while ($user && $usersScanned < self::USERS_PER_SESSION) {
 			$this->runScanner($user);
+			$user = $this->getUserToScan();
```
Comment (PVince81) [inline apps/files/lib/BackgroundJob/ScanFiles.php:122]: this also means that if errors occurred when scanning the previous user, we won't find the next one and will find the same user over and over again, which could cause an infinite loop, unless we are confident that scanning errors will always throw out

can you confirm this ?

File: lib/private/Files/Cache/Scanner.php [hunk 3]
```diff
@@ -499,19 +500,24 @@ public static function isPartialFile($file) {
 	 * walk over any folders that are not fully scanned yet and scan them
 	 */
 	public function backgroundScan() {
-		if (!$this->cache->inCache('')) {
-			$this->runBackgroundScanJob(function () {
-				$this->scan('', self::SCAN_RECURSIVE, self::REUSE_ETAG);
-			}, '');
+		if ($this->storage->instanceOfStorage(Jail::class)) {
+			$unJailedScanner = $this->storage->getUnjailedStorage()->getScanner();
```
Comment (PVince81) [inline lib/private/Files/Cache/Scanner.php:504]: in that cases do we need this ? is the background scanner also scanning into received shares and group folders ?

Comment (PVince81) [reply lib/private/Files/Cache/Scanner.php:504]: (please answer with a PHP comment if applicable)

File: lib/private/Files/Cache/Scanner.php [hunk 4]
```diff
@@ -499,19 +500,24 @@ public static function isPartialFile($file) {
...
-			while (($path = $this->cache->getIncomplete()) !== false && $path !== $lastPath) {
-				$this->runBackgroundScanJob(function () use ($path) {
-					$this->scan($path, self::SCAN_RECURSIVE_INCOMPLETE, self::REUSE_ETAG | self::REUSE_SIZE);
-				}, $path);
-				// FIXME: this won't proceed with the next item, needs revamping of getIncomplete()
-				// to make this possible
-				$lastPath = $path;
+			if (!$this->cache->inCache('')) {
+				$this->runBackgroundScanJob(function () {
+					$this->scan('', self::SCAN_RECURSIVE, self::REUSE_ETAG);
```
Comment (PVince81) [inline lib/private/Files/Cache/Scanner.php:509]: add comment to explain that we do this in case the user's filecache entries have not been initialized (ex: never logged in)

File: lib/private/Files/Utils/Scanner.php [hunk 5]
```diff
@@ -163,10 +163,6 @@ public function backgroundScan($dir) {
 				continue;
 			}
 
-			// don't scan received local shares, these can be scanned when scanning the owner's storage
-			if ($storage->instanceOfStorage(SharedStorage::class)) {
```
Comment (PVince81) [inline lib/private/Files/Utils/Scanner.php:167]: alright, so I guess we assume that the bkg scan will anyway only do a single pass through whoever received the shared storage and never have to touch it again, so it's ok to remove this condition for bkg scan (but keep it for regular scanner)

**Review by PVince81 — COMMENTED** [review]
Code looks good but I have a few concerns, please clarify by adding code comments where applicable

File: apps/files/lib/BackgroundJob/ScanFiles.php [hunk 6]
```diff
@@ -53,23 +52,20 @@ class ScanFiles extends \OC\BackgroundJob\TimedJob {
...
+		IConfig          $config,
 		IEventDispatcher $dispatcher,
-		ILogger $logger,
-		IDBConnection $connection
+		ILogger          $logger,
+		IDBConnection    $connection
 	) {
 		// Run once per 10 minutes
-		$this->setInterval(60 * 10);
+		$this->setInterval(1);
```
Comment (ChristophWurst) [inline apps/files/lib/BackgroundJob/ScanFiles.php:66]: every second? :see_no_evil:

Comment (icewind1991) [reply apps/files/lib/BackgroundJob/ScanFiles.php]: :speak_no_evil:

File: apps/files/lib/BackgroundJob/ScanFiles.php (hunk 2 above)
Comment (icewind1991) [reply apps/files/lib/BackgroundJob/ScanFiles.php:124]: The loop wouldn't have been infinite due to `USERS_PER_SESSION`, but I added an extra check to exit the loop early when we get stuck and log about it.

File: lib/private/Files/Utils/Scanner.php (hunk 5 above)
Comment (icewind1991) [reply lib/private/Files/Utils/Scanner.php:167]: Yes, the background scanner will now process the full source storage for shares

File: apps/files/lib/BackgroundJob/ScanFiles.php (hunk 1 above)
Comment (icewind1991) [reply apps/files/lib/BackgroundJob/ScanFiles.php:103]: Yes, the root is excluded because propagation doesn't always seem to update it's size correctly for home storages from my quick testing

**Comment by icewind1991** [comment]
ci happy and comments should be addressed

File: apps/workflowengine/lib/Check/FileSystemTags.php [hunk 7]
```diff
@@ -135,8 +135,8 @@ protected function getFileIds(ICache $cache, $path, $isExternalStorage) {
 		// TODO: Fix caching inside group folders
 		// Do not cache file ids inside group folders because multiple file ids might be mapped to
 		// the same combination of cache id + path.
-		$shouldCacheFileIds = !$this->storage
-			->instanceOfStorage(\OCA\GroupFolders\Mount\GroupFolderStorage::class);
+		/** @psalm-suppress InvalidArgument */
+		$shouldCacheFileIds = !$this->storage->instanceOfStorage(\OCA\GroupFolders\Mount\GroupFolderStorage::class);
```
Comment (PVince81) [inline apps/workflowengine/lib/Check/FileSystemTags.php:139]: looks unrelated

**Review by PVince81 — APPROVED** [review]
:+1: thanks a lot for the extensive code comments, it makes sense

it seems some formatting process went through the code, if it was all intended then good to go

**Comment by PVince81** [comment]
oh, a conflict. can you resolve ?

File: apps/workflowengine/lib/Check/FileSystemTags.php (hunk 7 above)
Comment (icewind1991) [reply apps/workflowengine/lib/Check/FileSystemTags.php:139]: It is needed due to the extra annotations added to `instanceOfStorage` and the fact that `GroupFolderStorage` doesn't exist when psalm is running

File: apps/files/lib/BackgroundJob/ScanFiles.php [hunk 8]
```diff
@@ -53,23 +51,20 @@ class ScanFiles extends \OC\BackgroundJob\TimedJob {
...
 	 * @param IDBConnection $connection
 	 */
 	public function __construct(
-		IConfig $config,
-		IUserManager $userManager,
+		IConfig          $config,
 		IEventDispatcher $dispatcher,
-		ILogger $logger,
-		IDBConnection $connection
+		ILogger          $logger,
```
Comment (CarlSchwan) [inline apps/files/lib/BackgroundJob/ScanFiles.php:61]: It looks like your formatter did some changes here. I'm not sure why php-cs-check didn't complain.

**Review by CarlSchwan — APPROVED** [review]
Looks good

**Comment by acsfer** [comment]
@icewind1991 currently on 22.2.3, background job for unscanned files is not reliable as randomly some folders gets unscanned and quotas are not calculated right; this is only fixed by `occ files:scan --all --unscanned`

Does this PR fix these cases or is this a different issue?
